flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironment.java

```diff
@@ -217,6 +218,7 @@ public ShuffleIOOwnerContext createShuffleIOOwnerContext(
 			for (InputGateDeploymentDescriptor igdd : inputGateDeploymentDescriptors) {
 				SingleInputGate inputGate = singleInputGateFactory.create(
 					ownerContext.getOwnerName(),
+					counter,
```

Comment:
Here, we rely on iteration order (`counter`), but the type of `inputGateDeploymentDescriptors` is `Collection`.
When it is created it is `List` though.
So I think we should change the type of `TaskDeploymentDescriptor.inputGates` and all parameters down to  `inputGateDeploymentDescriptors`  to `List`.

Reply:
I guess your suggestion is for unifying all the related data structures (collection, array) from `TaskDeploymentDescriptor.inputGates` to `List`?

Actually I do not see any correctness or performance problems ATM. This unification really spans cross multiple components from scheduler to task stack, and I do think it is really necessary to refactor it now. Or if we want to do that, it should be a separate ticket out of this PR scope. We only make use of previous `counter` to give a correct gate index while creating in iteration for this PR motivation. WDYT?

Reply:
Not sure if I got you right.

I meant to change the type from `Collection` to `List` for:
1. `inputGateDeploymentDescriptors` parameter in `NettyShuffleEnvironment#createInputGates`
1. `TaskDeploymentDescriptor#inputGates` field
1. `inputGateDeploymentDescriptors` parameter in `Task` constructor
1. `inputGateDeploymentDescriptors` parameter in `TaskDeploymentDescriptor` constructor

Without it, it's easy to break the ordering, at least for me :)

I think it should be the same PR because requirement for ordering was added here.

Reply:
I thought you were referring to the unnecessary transformation in the logics before, sorry for misunderstanding.

You are right that the current `Collection` actually does not provide any order guarantee which relies on the subclass interface (`List` or `Set`), and we indeed need the determined order after restarting.

It is reasonable to adjust the `Collection` to `List` as you suggested, but the only trouble is that it would change the public interface of `ShuffleEnvironment` proposed in FLIP-31. So it might break the compatibility if some outside users already relied on it. I guess it needs to discussion in ML for finalization.  Anyway, I already addressed this issue in a separate commit.